Benchmarking working_team

conectado opened the pull request:
Pull Request to track progress in #1534 
---

We need to benchmark `working_team` to implement the weights for each extrinsic(and `on_initialize`)

In order to test the benchmark, I added a dummy implementation of `working_team::Trait` for the `Runtime` we are using.

**Note**: It's useless to calculate the weights in the reference hardware until the acual impl for the `Runtime` is done. (Benchmarking only make sense in the context of a given `Runtime`)

To complete the PR I need to do the following:

- [x] Add benchmarking to the Runtime(when the feature is enabled)
- [x] Add dummy `working_team::Trait` impl for `Runtime`
- [x] Add benchmark function for each branch of each extrinsic and `on_initialize`
  - [x] `leave_role`
  - [x] `set_budget`
  - [x] `add_opening`
  - [x] `update_reward_account`
  - [x] `set_status_text`
  - [x] `update_reward_amount`
  - [x] `spend_from_budget`
  - [x] `decrease_stake`
  - [x] `increase_stake`
  - [x] `terminate_role`
  - [x] `slash_stake`
  - [x] `withdraw_application`
  - [x] `cancel_opening`
  - [x] `update_role_account`
  - [x] `fill_opening`
  - [x] `apply_on_opening`
  - [x]  `on_initialize`
- [x] Add code to run verifications with the tests
- [x] Implement the verify function for each benchmark function
- [x] Whitelist appropiate keys
- [ ] Write the weight function for each extrinsic (**Note**:  in substrate current version this is done automatically by the CLI) 
- [ ] Run the code in [reference hardware](https://substrate.dev/docs/en/knowledgebase/learn-substrate/weight) (**Blocked until `working_team::Trait` is implemented**)
- [ ] Add weights in the extrinsics(and `on_initialize`)

conectado committed:
benchmark: working-team: add mock runtime impl and benchmark for leave_role

* Adds a mock impl for working_team in the runtime needed to test benchmark code
* Adds benchmarks in the runtime
* Adds benchmark for working_team
* Implement first draft for leave_role benchmarking implementation

conectado committed:
benchmark: working_team: final version of leave_role

* Update leave_role for a more finalized version with all path tested
* Update `StakingHandler` to be able to test `leave_role` worst case

bedeho commented:
What do you think is the best way to review this?

Should this all be reviewed when this is all done, in one gigantic review? or something else?

conectado commented:
> What do you think is the best way to review this?
> 
> Should this all be reviewed when this is all done, in one gigantic review? or something else?

I thought it over and I think the best way to approach and I think the best way is to do a review each time a certain "milestone" is hit. That way we can catch problems early and don't have to re-write a lot of code, but not so often that you review code that will be discarded or changed later.

I was thinking of these milestones.
When each one of these are completed:

- [x] Add benchmark function for each branch of each extrinsic and `on_initialize`
- [ ] Implement the verify function for each benchmark function
- [ ] Whitelist appropiate keys
- [ ] Everything

Does this make sense?

conectado committed:
benchmark: working_team: add 4 extrinsics

This commit adds the following extrinsics:
* `set_budget`
* `add_opening`
* `update_reward_account`
* `set_status_text`

conectado committed:
benchmark: working-team: use add/apply/fill-opening

Updates benchmarks for working to use add/apply/fill extrinsics instead of
directly modifying the storage

conectado committed:
runtime: benchmark: use a non-lead when it is worst add extrinsics

This commit refactors the benchmark to use a non-lead in some cases where the checks make it a worst case scenario

Furthermore, it adds the following extrinsic benchmarks:
* `update_reward_amount`
* `spend_from_budget`
* `decrease_stake`

conectado committed:
benchmark: working-team: add unpaid reward for workers and extrinsics

This commits adds an updaid reward for each created worker

This is the worst case scenario for all the extrinsics treated until now

Also adds extrinsics:
* `increase_stake`
* `terminate_role`
* `slash_stake`

conectado committed:
benchmark: working-team: add benchmark for 2 extrinsics

The added benchmarks are for the extrinsics:
* withdraw_application
* cancel_opening

conectado committed:
benchmark: working-team: add benchmark for update_role_account

conectado committed:
benchmark: working-team: refactor `insert_a_worker` and add benchmark extrinsic

This commit refactor `insert_a_worker` for clarity and add the benchmark
for the extrinsic `fill_opening`.

conectado committed:
benchmark: working-team: implement apply_on_opening benchmark

This commit adds the benchmark for apply_on_opening extrinsic

conectado committed:
benchark: working-team: adds benchmarking for `on_initialize`

conectado committed:
benchmark: working-team: refactors  small changes

conectado committed:
benchmark: working-team: refactor extrinsics add/apply opening

conectado committed:
benchmark: working-team: refactor on_initialize

conectado committed:
benchmark: working-team: add on_initialize benchmark branch and refactor

conectado commented:
> > What do you think is the best way to review this?
> > Should this all be reviewed when this is all done, in one gigantic review? or something else?
> 
> I thought it over and I think the best way to approach and I think the best way is to do a review each time a certain "milestone" is hit. That way we can catch problems early and don't have to re-write a lot of code, but not so often that you review code that will be discarded or changed later.
> 
> I was thinking of these milestones.
> When each one of these are completed:
> 
>     * [x]  Add benchmark function for each branch of each extrinsic and `on_initialize`
> 
>     * [ ]  Implement the verify function for each benchmark function
> 
>     * [ ]  Whitelist appropiate keys
> 
>     * [ ]  Everything
> 
> 
> Does this make sense?

The extrinsic/`on_initialize` is ready to be reviewed, a couple of pointers:

* I added a dummy implementation for the `working_team::Trait` that was needed to test the benchmarks
  * This meant that I also needed to implement a dummy implementation of `working_team::StakingHandler`
    * `current_stake` returns a value different than 0, this is needed for `leave_role_later` to take the correct branch
  * I also needed to expose publicly `types::BalanceOfCurrency` to implement this dummy quickly
* **It's very important that we check that every extrinsic benchmarked takes the worst possible scenario or when it's not obvious every not-obviously worst-case is tested**
  * I always erred in the side of caution trying to test every branch when I wasn't sure
* **It's also very important to see that we identify every parameter which makes the function takes more time** (such as the length of a string that we are hashing or the number of workers over which we are iterating)
  * For `on_initialize` it might appear that there are 2 parameters: the number of leaving workers and the number of workers to reward. However, when there are two parameters when the benchmarks do its linear regression it fixes the other parameter in the biggest case, in this case `T::MaxWorkerNumberLimit::get()`, this means that there is no space for leaving workers because all the workers would be in the rewarded set or vice versa. Still, we can take this as 2 branches with only 1 parameter, this being `leaving_workers` + `rewarding_workers`(with the two branches being all are leaving or all rewarded). This works because one of the branches has to be worse than the other and the more workers taking the "bad" branch the worse it will be.
  * Some of the parameters seem to be arbitrarily going from 1 to 10, these are the constant cases but a parameter is needed for the benchmarking to work(afaik in the latest substrate version this is no longer needed)
  * Some parameters go into an arbitrarily big number, these are some input parameters that were unbounded, in every example I've found every parameter was bounded and used that as the upper limit. As I understand the weight of the parameter would prevent this to be a problem, but I'm not sure, we might have to look more deeply into that.

conectado commented:
For whitelisting there are two cases:
* A key is read/written in every block 
* A key is transient

In both cases the since the key is cached and we don't need to count it on our read/write count

With the version we are using for Substrate we can't add a pallet scoped key because the function to do this(`add_to_whitelist`) was added in [this commit](https://github.com/paritytech/substrate/commit/368903f7aa9ef652bf8157d476dc13cb36e3affc#diff-0531fde3df80cc3c15bc09edd247ac8cbd6106ec9adfd406a49a07528ffea663R126) which was merged after the version we are using. (This shouldn't matter since we care more about using the worst case scenario rather than tightening the weight to the actual computation time)

Still, in this case I think we shouldn't whitelist any key regarding this pallet, since the only candidates are the keys in `WorkerById` which are accessed through the function `Self::get_workers_with_finished_unstaking_period`. However then we can add and access another key in `WorkerById` in the same block, so we have no way of knowing a-priori if the key in the current extrinsic was previously accessed.

For these reasons, I'm marking the checkpoint **Whitelist appropiate keys** as done.

conectado commented:
Note: To implement the benchmark tests we need to use `DefaultInstance` as explained in #1652

conectado committed:
benchmark: working-team: update working-team to run benchmark verify

The changes to achieve this are:
* Use `DefaultInstance` also use it in the tests
* Use `add_screned_member` instead of `buy_membership`
* Enable benchmarking tests

shamil-gadelshin commented on runtime-modules/working-team/Cargo.toml:
```diff
@@ -35,4 +37,5 @@ std = [
 	'common/std',
 	'membership/std',
 	'balances/std',
```
Do we need 'frame-benchmarking' in the std?

conectado replied:
I think we don't need it. It is a leftover from earlier tests.

Will erase it.

shamil-gadelshin commented on runtime/src/lib.rs:
```diff
@@ -65,6 +65,95 @@ pub use versioned_store;
 pub use versioned_store_permissions;
 pub use working_group;
 
```
We probably need to remove this after the benchmarking is done and prior to merging.

shamil-gadelshin commented on runtime-modules/working-team/src/lib.rs:
```diff
@@ -47,7 +48,12 @@ use sp_std::vec::Vec;
 
 pub use errors::Error;
-use types::{ApplicationInfo, BalanceOfCurrency, MemberId, TeamWorker, TeamWorkerId, WorkerInfo};
+/*
```
We probably need to remove this after the benchmarking is done and prior to merging.

shamil-gadelshin commented on runtime/src/lib.rs:
```diff
@@ -618,5 +707,9 @@ construct_runtime!(
         // --- Working groups
         // reserved for the future use: ForumWorkingGroup: working_group::<Instance1>::{Module, Call, Storage, Event<T>},
         StorageWorkingGroup: working_group::<Instance2>::{Module, Call, Storage, Config<T>, Event<T>},
+        /*
```
We probably need to remove this after the benchmarking is done and prior to merging.

shamil-gadelshin commented on runtime/src/runtime_api.rs:
```diff
@@ -8,6 +8,8 @@ use sp_core::crypto::KeyTypeId;
 use sp_core::OpaqueMetadata;
 use sp_runtime::traits::{BlakeTwo256, Block as BlockT, NumberFor};
 use sp_runtime::{generic, ApplyExtrinsicResult};
+#[cfg(feature = "runtime-benchmarks")]
```
We could move this inside the `impl frame_benchmarking::Benchmark<Block> for Runtime {` similar to other `use` statements.

shamil-gadelshin commented on runtime/src/runtime_api.rs:
```diff
@@ -201,4 +203,53 @@ impl_runtime_apis! {
+
+            let whitelist: Vec<Vec<u8>> = vec![
+                // Block Number
+                // frame_system::Number::<Runtime>::hashed_key().to_vec(),
```
Remove the commented code.

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,599 @@
+    T::ApplicationId::from(id.try_into().unwrap())
+}
+
+fn add_opening_and_n_apply<T: Trait<I>, I: Instance>(
```
What does `n` correspond to? Please rename it.

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,599 @@
+    }: { WorkingTeam::<T, _>::on_initialize(curr_block_number) }
+
+    apply_on_opening {
+      let i in 1 .. 50000;
```
We should make the max description length consistent with the UI restrictions or limit this value in the runtime.

conectado replied:
I believe just like in `add_opening` we need to restrict this value in the runtime. (Will add a similar comment)

bedeho replied:
What about this

https://github.com/Joystream/joystream/pull/1587#discussion_r520490279

conectado committed:
benchmark: working-team: add verify function for each benchmark

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,599 @@
+      successful_application_ids.insert(application_id);
+    }: fill_opening(RawOrigin::Root, opening_id, successful_application_ids)
+
+    fill_opening_worker { // We can actually fill an opening with 0 applications?
```
> We can actually fill an opening with 0 applications?

Yes, this is a bug. I will fix it myself during the pallet integration.

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,599 @@
+    }: _ (RawOrigin::Signed(lead_account_id.clone()), apply_on_opening_params)
+
+    fill_opening_lead {
+      let i in 0 .. 10;
```
Do we need "i" variable here?

shamil-gadelshin replied:
I saw your comment `// TODO: test not running if we don't set a range of values`. If this is really the case, please, add the explicit comment near each usage or a single one in the module description.

conectado replied:
This is needed in Substrate 2.0.0-RC-4 in the latest version(2.0.0) it seems that it is no longer needed. Will change this to `let i in 0 ..  1` and a comment to erase it when we update the version.

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,599 @@
+    }: leave_role(RawOrigin::Signed(caller_id), caller_worker_id)
+    verify { assert_eq!(true, true); }
+}
+
```
Yes, please add tests for each benchmark functions.

shamil-gadelshin replied:
Nevermind, the latest commit includes the tests.

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,599 @@
+      let (lead_id, lead_worker_id) = insert_a_worker::<T, I>(StakingRole::WithoutStakes, JobOpeningType::Leader, 0, None);
+      let (caller_id, worker_id) = insert_a_worker::<T, I>(StakingRole::WithStakes, JobOpeningType::Regular, 1, Some(lead_id.clone()));
+      let penalty = Penalty {
+        slashing_text: vec![],
```
Should we place a major vector here?

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,599 @@
+      // To be able to pay unpaid reward
+      WorkingTeam::<T, _>::set_budget(RawOrigin::Root.into(), current_budget).unwrap();
+      let penalty = Penalty {
+        slashing_text: vec![],
```
Should we place a major vector here?

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,599 @@
+    // Regular opening is the worst case scenario since the checks
+    // require access to the storage whilist that's not the case with a lead opening
+    add_opening{
+      let i in 0 .. 50000; // TODO: We should have a bounded value for description
```
@bedeho Despite the fact we don't use the text fields in the runtime because of the query node - we need to bound the length of these values in order to properly benchmark such processes like text hashing.

bedeho replied:
I think there is a way to do this without needing to bound it, I believe the extrinsic caller needs to commit to a length. Please elaborate @conectado.

conectado replied:
That way of doing things works for weights that depend on an stored item.

Say, for example, an extrinsic that depends on the number of stored workers, then instead of assuming worst-case scenario(that there are max number of workers stored) you can ask the caller to pass the number of workers as a parameter and then ensure that it is correct before executing the code. That way you can use that parameter to calculate the weight instead of assuming worst-case and refunding later.

Here, it is not the case, because we depend directly on the parameter passed, the problem is that this parameter is unbounded and for what I've seen in FRAME is that they actually bound with a "max" value all these linear parameters (like [here](https://github.com/paritytech/substrate/blob/900d034fc1ba0ccb30dd54b2dd7498ad5caa7fb5/frame/treasury/src/lib.rs#L633)).

However, I don't fully understand why this is needed. What I understand is that since we are doing linear interpolation, if the extrinsic (meaning, in this case, the hash function) is really linear we can use an arbitrarily large value as an upper-bound only for the linear interpolation and then depend on the Weight of the extrinsic call to bound the function running-time. But I've not seen it done like that anywhere.

shamil-gadelshin replied:
Ok. Let's skip the upper limits and rely on block length fees and linear dependency on the text length.

shamil-gadelshin reviewed (changes requested):
- Please, refactor the code in order to be closer to rust recommended 99 symbols per line.
- Did you consider 'weight-refunds' for any function?

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,921 @@
+      for i in 1..successful_application_ids.len() {
+        assert!(WorkerById::<T, I>::contains_key(TeamWorkerId::<T>::from(i.try_into().unwrap())), "Not all workers
+        added");
+      }
```
It seems that you missed check for the `OpeningFilled` event.

shamil-gadelshin replied:
Also, a successful outcome contains a new worker in the storage.

conectado replied:
> Also, a successful outcome contains a new worker in the storage.

Aren't we checking that the outcome contains a new worker in the storage by checking that:
```rs
WorkerById::<T, I>::contains_key(TeamWorkerId::<T>::from(i.try_into().unwrap()))
```

is true?

shamil-gadelshin replied:
Nevermind, probably I saw not the latest commit version.

shamil-gadelshin commented on runtime-modules/working-team/src/benchmarking.rs:
```diff
@@ -0,0 +1,921 @@
+    // Regular worker is the worst case scenario since the checks
+    // require access to the storage whilist that's not the case with a lead opening
+    increase_stake {
+      let i in 0 .. 10;
```
For stake related 'verify' functions, you could check the actual change of the stake.

conectado replied:
I think a problem with that it's that it depends on the specific implementation of `StakingHandler` in the runtime or in the verification test in the `Test`.

In fact, I think we can't check for `current_stake` since that (almost)always returns 100  in the current mock implementation. I'm not sure but maybe we can for the locked amount?

shamil-gadelshin replied:
I agree. Let's skip it then.

shamil-gadelshin commented:
Please, introduce the '4 spaces' indentation instead of the '2 spaces'.

shamil-gadelshin commented:
Also, please avoid double line spacing like in the `add_opening` or `on_initialize_leaving` and similar places.

conectado committed:
benchmark: working-team: fix formatting errors

conectado committed:
benchmark: working-team: small refactor and benchmark on penalty text

bedeho reviewed (commented):
Very impressive. I don't have the ability to look through this in any detail, but before any non-trivial changes are implemented in the working group module to accommodate the benchmarking, lets schedule a shared call going through the major issues, as I assume the initial set of issues will be reoccurring across the runtime.

conectado commented:
>     * Did you consider 'weight-refunds' for any function?

`leave_role` can have weight returns depending on the branch(leaving immediatly or later) taken and we could implement  weight refunds for `terminate_role` in a similar fashion, depending if the terminated role is a lead or a worker, the same goes for `fill_opening`.

What do you think?

shamil-gadelshin commented:
Please, remove all TODO code from the PR before merging. We are going to create an integration branch and this mock code in the runtime is redundant. 

Regarding the 'weight-refunds' - I was thinking about the major difference between code branches weights and which is hidden in the subtle 'parameters and storage state' combination. Since there are no such obvious places - we should wait for production weight costs of extrinsic and consider 'weight-refunds' as an advanced tweaking technique.

As to leave_role, termination_role and similar: you had already split their benchmarks into several functions and I think we could use just max of them because the weight difference which brings the 'leader case' is not very significant.

shamil-gadelshin commented:
Also, since we are not going to limit the text fields, please, remove the comments in the benchmarks about it.

shamil-gadelshin reviewed (changes requested):
See the comments above.

conectado committed:
benchmark: working-team: remove todo comments

conectado committed:
benchmark: working-team: revet changes for local benchmark

shamil-gadelshin reviewed (approved):
LGTM

conectado committed:
benchmark: working-team: update travis to use needed chain for benchmarks

conectado committed:
benchmark: working-team: update version

conectado committed:
benchmark: working-team: remove unnecesary comment

conectado committed:
benchmark: working-team: update `handle_from_id` function
